core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooSession.java

```diff
@@ -119,8 +119,19 @@ public static ZooKeeper connect(String host, int timeout, String scheme, byte[]
     boolean tryAgain = true;
     long sleepTime = 100;
     ZooKeeper zooKeeper = null;
-
-    long startTime = System.currentTimeMillis();
+    /*
+     * Originally, startTime = System.currentTimeMillis(). Changed to System.nanoTime() to more
+     * accurately compute durations because it is not based on system clock variations. The
+     * ZooKeeper method signature expects an int value for 'timeout' and performs several
+     * calculations that can result in Numeric Expression Overflow errors if large, nanosecond
+     * values are used. For ths reason, System.nanoTime() is is converted to MS units prior to being
+     * used in calculations. Although, the resolution of 'startTime' is still in the MS range, the
+     * value from which it is calculated is nanoTime. Also, the MS units are preserved in the
+     * original method code, without the need for conversion. TimeUnit.convert is not used because
+     * "conversions from fine to coarser granularities truncate, so lose precision" (Oracle Java 7
+     * API)
+     */
+    long startTime = System.nanoTime() / 1000000;
```

Comment:
It's not a good idea to divide here, since it is the differences between two nano-times that matter, and you are losing precision here in a way that could affect the subtraction later.

The correct pattern to do the conversion to millis would be something like this:

```java
long startTime = System.nanoTime();
// code here...
long laterTime = System.nanoTime();
long timeDurationMillis = TimeUnit.NANOSECONDS.toMillis(laterTime - startTime);
```

It is okay to use the TimeUnit conversions here, and lose precision, because it's okay to lose the precision in the difference. It is not okay to lose precision on the values before computing the difference in times. (Also, using this method to convert is the same as dividing... they both lose precision in the same way.)

Reply:
Indeed, that makes more sense. Working the issue along those lines, there are two sections in ZooSession where calculations are performed using millis values for all the operands:

`if (System.currentTimeMillis() - startTime > 2L * timeout) {
        throw new RuntimeException("Failed to connect to zookeeper (" + host
            + ") within 2x zookeeper timeout period " + timeout);
      }`

`if (tryAgain) {
        if (startTime + 2L * timeout < System.currentTimeMillis() + sleepTime + connectTimeWait)
          sleepTime = startTime + 2L * timeout - System.currentTimeMillis() - connectTimeWait;
        if (sleepTime < 0) {
          connectTimeWait -= sleepTime;
          sleepTime = 0;
        }
        UtilWaitThread.sleep(sleepTime);
        if (sleepTime < 10000)
          sleepTime = sleepTime + (long) (sleepTime * secureRandom.nextDouble());
      }`

One solution is to convert from millis to nano when the values are first assigned. There are a couple issues I see with doing that. One, timeout is still being passed into the method in millis and would need to be converted at the point-of-use (or assigned a value within the method, which breaks convention). Two, there are calculations that use these values, but do not use startTime, that are already working correctly. Also, TIME_BETWEEN_CONNECT_CHECKS_MS would likely need to be refactored to ...-NS.

Another solution (pasted below) is to  do conversions on the operands. This will prevent the Numeric Overflow Errors, but kind of decreases the readability of the code. Scaling up the operand values might not make them more accurate, but we should not _lose_ accuracy this way.

`if (TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime) > 2L * timeout) {
        throw new RuntimeException("Failed to connect to zookeeper (" + host
            + ") within 2x zookeeper timeout period " + timeout);
      }`
`if (tryAgain) {
        if (TimeUnit.NANOSECONDS.toMillis(startTime + 2L * TimeUnit.MILLISECONDS.toNanos(timeout))
            < TimeUnit.NANOSECONDS
                .toMillis(System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(sleepTime)
                    + TimeUnit.MILLISECONDS.toNanos(connectTimeWait))) {`
`sleepTime =
              TimeUnit.NANOSECONDS.toMillis(startTime + 2L * TimeUnit.MILLISECONDS.toNanos(timeout)
                  - System.nanoTime() - TimeUnit.MILLISECONDS.toNanos(connectTimeWait));
        }`

Reply:
@cradal I updated your comment's Markdown to use syntax highlighting, to make it easier to read, but I think it'd be easier if you just updated the PR with the new changes you are proposing, rather than paste it in to a comment. It will be easier to review the proposed changes inline, rather than in a discussion thread. Also, if you want to make your code more readable, you could statically import TimeUnit.NANOSECONDS and TimeUnit.MILLISECONDS.